src/raftstore/coprocessor/region_collection.rs

```diff
@@ -0,0 +1,713 @@
+
+/// `RegionCollection` has its own thread (namely RegionCollectionWorker). Queries and updates are
+/// done by sending commands to the thread.
+enum RegionCollectionMsg {
+    RaftStoreEvent(RaftStoreEvent),
+    /// Get all contents from the collection. Only used for testing.
+    DebugDump(mpsc::Sender<(RegionsMap, RegionRangesMap)>),
+}
+
+impl Display for RegionCollectionMsg {
+    fn fmt(&self, f: &mut Formatter) -> FmtResult {
+        match self {
+            RegionCollectionMsg::RaftStoreEvent(e) => write!(f, "RaftStoreEvent({:?})", e),
+            RegionCollectionMsg::DebugDump(_) => write!(f, "DebugDump"),
+        }
+    }
+}
+
+/// `EventSender` implements observer traits. It simply send the events that we are interested in
+/// through the `scheduler`.
+#[derive(Clone)]
+struct EventSender {
+    scheduler: Scheduler<RegionCollectionMsg>,
+}
+
+impl Coprocessor for EventSender {}
+
+impl RegionChangeObserver for EventSender {
+    fn on_region_changed(&self, context: &mut ObserverContext, event: RegionChangeEvent) {
+        let region = context.region().clone();
+        let event = match event {
+            RegionChangeEvent::Create => RaftStoreEvent::CreateRegion { region },
+            RegionChangeEvent::Update => RaftStoreEvent::UpdateRegion { region },
+            RegionChangeEvent::Destroy => RaftStoreEvent::DestroyRegion { region },
+        };
+        self.scheduler
+            .schedule(RegionCollectionMsg::RaftStoreEvent(event))
+            .unwrap();
+    }
+}
+
+impl RoleObserver for EventSender {
+    fn on_role_change(&self, context: &mut ObserverContext, role: StateRole) {
+        let region = context.region().clone();
+        let event = RaftStoreEvent::RoleChange { region, role };
+        self.scheduler
+            .schedule(RegionCollectionMsg::RaftStoreEvent(event))
+            .unwrap();
+    }
+}
+
+/// Create an `EventSender` and register it to given coprocessor host.
+fn register_raftstore_event_sender(
+    host: &mut CoprocessorHost,
+    scheduler: Scheduler<RegionCollectionMsg>,
+) {
+    let event_sender = EventSender { scheduler };
+
+    host.registry
+        .register_role_observer(1, box event_sender.clone());
+    host.registry
+        .register_region_change_observer(1, box event_sender.clone());
+}
+
+/// `RegionCollectionWorker` is the underlying runner of `RegionCollection`. It listens on events
+/// sent by the `EventSender` and maintains the collection of all regions. Role of each region
+/// are also tracked.
+struct RegionCollectionWorker {
+    // region_id -> (Region, State)
+    regions: HashMap<u64, RegionInfo>,
+    // 'z' + end_key -> region_id
+    region_ranges: BTreeMap<Vec<u8>, u64>,
+}
+
+impl RegionCollectionWorker {
+    fn new() -> Self {
+        Self {
+            regions: HashMap::default(),
+            region_ranges: BTreeMap::default(),
+        }
+    }
+
+    /// Check if the end_key has collided with another region. If so, try to clean it.
```

Comment:
I think you should describe the handling logic for region split and region merge.